Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp

```diff
@@ -1342,14 +1345,26 @@ std::vector<double> DiscusMultipleScatteringCorrection::simulatePaths(
       std::transform(weights.begin(), weights.end(), sumOfWeights.begin(), sumOfWeights.begin(), std::plus<double>());
       std::transform(weights.begin(), weights.end(), countZeroWeights.begin(), countZeroWeights.begin(),
                      [](double d, int count) { return d > 0. ? count : count + 1; });
+
+      // increment standard deviation using Welford algorithm
+      for (size_t i = 0; i < wValues.size(); i++) {
+        deltas[i] = weights[i] - weightsMeans[i];
+        weightsMeans[i] += deltas[i] / static_cast<double>(ie + 1);
+        weightsM2[i] += deltas[i] * (weights[i] - weightsMeans[i]);
+        // calculate sample SD (M2/n-1)
+        // will give NaN for m_events=1, but that's correct
+        weightsErrors[i] = sqrt(weightsM2[i] / static_cast<double>(ie));
```

Comment:
I think the variance doesn't need to be divided by `ie` here (the mean is updated by the deviation weighted by `ie+1` and then the product of the deviation from the old mean and the deviation from the new mean is added to the variance sum.
At a later point you divide by `sqrt(N-1)` in another loop, so maybe it's a typo and really you should divide by `sqrt(nPaths)` inside this loop and ditch the later loop?

Reply:
I've had a think about this. I think the division by `ie` on L1356 corresponds to the "M2 / count" here from the Wikipedia page that I've highlighted in yellow:
![image](https://user-images.githubusercontent.com/56295817/208728628-dd0f5630-95d8-463e-b074-e0500bf7e29d.png)
This sets weightsErrors[i] equal to the standard deviation of the weights for a particular omega value. I probably don't need to do L1356 on every iteration of the loop - each execution of L1356 just overwrites the previous answer but the repetition is harmless. **Note** - at this point weightsErrors[i] is just a standard deviation and not an error. The naming of the vector here isn't great because it's based on what the vector will hold at the end of the function.

I then need to do a further division by sqrt(nPaths) on L1364 to get the error on the mean weight for a particular omega value. This bit isn't on the Wikipedia page at all because the Welford algorithm is all about getting the variance\SD of a set of values. It doesn't cover getting the error on the mean. Does that make sense?

As discussed this is perhaps easier to see in MCAbsorptionStrategy.cpp. I tried removing the final division in MCAbsorptionStrategy.cpp and running the MonteCarloAbsorption algorithm using this script and the errors on the abscor workspace seemed too large which gives me a bit of confidence my argument above is valid:

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

data = CreateSampleWorkspace(WorkspaceType='Histogram', NumBanks=1)
data = ConvertUnits(data, Target="Wavelength")
# Default up axis is Y
SetSample(data, Geometry={'Shape': 'Cylinder', 'Height': 5.0, 'Radius': 1.0,
                  'Center': [0.0,0.0,0.0]},
                Material={'ChemicalFormula': '(Li7)2-C-H4-N-Cl6', 'SampleNumberDensity': 0.07})
# Simulating every data point can be slow. Use a smaller set and interpolate
abscor = MonteCarloAbsorption(InputWorkspace=data,EventsPerPoint=1000)
corrected = data/abscor
```

Reply:
I think there is a bug, to show it I converted the code to python - so I may have made an error, but I think this is what the code is doing (converted into python)
```
import numpy as np

avg = 1
std = 0.2

nomega = 5
npaths = 1000
weights = np.random.normal(loc=avg, scale=std, size=(npaths, nomega))
# calculate the avg and std for each omega
print('avg = ', np.mean(weights, axis=0))
print('std = ', np.std(weights, axis=0))
# avg =  [0.9997892  0.99953238 0.99211105 1.01015046 1.00047634]
# std =  [0.19445069 0.20230386 0.20706431 0.20533679 0.19615291]

# use welford one-pass as in DiscusMultipleScatteringCorrection
deltas = np.zeros(nomega)
weightsMeans = np.zeros(nomega)
weightsM2 = np.zeros(nomega)
weightsErrors = np.zeros(nomega)
for ipath in range(npaths):
    for iomega in range(nomega):
        deltas[iomega] = weights[ipath, iomega] - weightsMeans[iomega]
        weightsMeans[iomega] += deltas[iomega] / (ipath + 1)
        weightsM2[iomega] += deltas[iomega] * (weights[ipath, iomega] - weightsMeans[iomega])
        weightsErrors[iomega] = np.sqrt(weightsM2[iomega] / ipath)
# at this point weightsErrors = [0.19454799 0.20240509 0.20716792 0.20543954 0.19625106]
# roughly consistent with std
# but then there is an additional division by sqrt(nPaths)
for iomega in range(nomega):
    weightsErrors[iomega] = weightsErrors[iomega] / np.sqrt(npaths);
  
print(weightsMeans)  # [0.9997892  0.99953238 0.99211105 1.01015046 1.00047634]
print(weightsErrors) # [0.00615215 0.00640061 0.00655122 0.00649657 0.006206  ]
```
The means agree with `np.mean` but I think the additional division by `sqrt(npaths)` is not needed - before that the values of `weightsErrors` were roughly consistent with `np.std` but there is still a discrepancy.

I think this is because in the first set of loops you divide weightErrors[iomega] by `'ipath`  and I think it should be `npaths`
```
# I propose it should be  this
deltas = np.zeros(nomega)
weightsMeans = np.zeros(nomega)
weightsM2 = np.zeros(nomega)
weightsErrors = np.zeros(nomega)
for ipath in range(npaths):
    for iomega in range(nomega):
        deltas[iomega] = weights[ipath, iomega] - weightsMeans[iomega]
        weightsMeans[iomega] += deltas[iomega] / (ipath + 1)
        weightsM2[iomega] += deltas[iomega] * (weights[ipath, iomega] - weightsMeans[iomega])
        weightsErrors[iomega] = np.sqrt(weightsM2[iomega] / npaths)
print(weightsMeans)  # [0.9997892  0.99953238 0.99211105 1.01015046 1.00047634]
print(weightsErrors) # [0.19445069 0.20230386 0.20706431 0.20533679 0.19615291]
```
Now the results agree with `np.std`
But maybe I've translated it to python wrong...

Reply:
I'm not trying to calculate np.std here though. I'm trying to calculate the error on the mean which is np.std/sqrt(n)

Reply:
Ah the standard error, that's OK then! I still think there will be a small inaccuracy because you divide by `ie` in the loop instead of `nPaths` - actually it is probably better to divide by the number of successful paths (which will have to wait after the loop over nPaths)

Reply:
There is a retry in the `ie` loop that means the final value of `ie` will always be `nPaths-1` when the loop completes. As discussed there is in theory a chance the retry mechanism will cause the loop to never complete but I've never seen this happen. I've deliberately divided by `nPaths-1` here because that gives a better estimate (unbiased) of the standard deviation than dividing by `nPaths`.